```{r setup, message=FALSE, warning=FALSE}
library(GenomicRanges)
library(ggpubr)
library(viridis)
library(ggbeeswarm)
library(tidyverse)
ROOT_DIR <- "/mnt/isilon/sfgi/dudekm/"
```

```{r}
FP_METHOD <- "PRINT_beta_cloglog"
dir.create("figures/%s" %>% sprintf(FP_METHOD), showWarnings = FALSE)
```

```{r}
regression_results <- readRDS(sprintf("../../regression/FP_methods/%s/regression_results/fp_score_covariates_genotype_regression.Rds", FP_METHOD))

Q_THRESHOLD <- 0.05

regression_results %>%
  mutate(fpQTL = ST_qval < Q_THRESHOLD) ->
  regression_results

# For beta regression, join with linear effect sizes
if (grepl("PRINT_beta", FP_METHOD, fixed = TRUE)) {
  linear_regression <- readRDS("../../regression/FP_methods/PRINT_no_gaussian/regression_results/fp_score_covariates_genotype_regression.Rds")
  
  regression_results$linear_beta <- linear_regression$beta
}
```

Load TSS ranges from RefSeq Curated
```{r}
refSeqCurated <- read.table(paste0(ROOT_DIR, "raw_data/genome/RefSeq/ncbiRefSeqCurated.txt"))

colnames(refSeqCurated) <- c("bin", "name", "chrom", "strand", "txStart", "txEnd", "cdsStart", "cdsEnd", "exonCount", "exonStarts", "exonEnds", "score", "name2", "cdsStartStat", "cdsEndStat", "exonFrames")

refSeqCurated %>%
  mutate(
    tx = ifelse(strand == "+", txStart, txEnd)
  ) ->
  refSeqCurated

```

```{r}
variant.range <- GRanges(seqnames = regression_results$variant_chrom, 
                         ranges = IRanges(start = regression_results$variant_pos, 
                                          end = regression_results$variant_pos))

tss.range <- GRanges(seqnames = refSeqCurated$chrom, 
                     ranges = IRanges(start = refSeqCurated$tx, 
                                      end = refSeqCurated$tx,
                                      names = refSeqCurated$name2))
```

```{r}
result <- distanceToNearest(variant.range, tss.range)
nearest_index <- subjectHits(result)
regression_results$nearest_tss <- refSeqCurated$name2[nearest_index]
regression_results$distance_to_nearest_tss <- mcols(result)$distance
regression_results$promoter <- regression_results$distance_to_nearest_tss <= 1000
```

```{r}
regression_results$distance_to_nearest_tss %>% quantile(probs = seq(0, 1, by = .2))
```

```{r fig.width=4, fig.height=4}
distance_summary <- regression_results %>%
  mutate(distance_bin = cut(distance_to_nearest_tss, 
                            breaks = c(0, 1000, 10000, 25000, 50000, 100000, Inf),
                            ordered_result = TRUE, right = FALSE)) %>%
  group_by(distance_bin) %>%
  summarize(
    n = n(),
    n_fpQTLs = sum(fpQTL)
  ) %>% 
  rowwise() %>%
  mutate(
    prop_fpQTLs = n_fpQTLs / n,
    prop_fpQTLs_95low = prop.test(n_fpQTLs, n)$conf.int[1],
    prop_fpQTLs_95high = prop.test(n_fpQTLs, n)$conf.int[2]
  )
```

# MANUSCRIPT - Figure 2C
```{r fig.width=3, fig.height=3}
distance_summary %>%
  ggplot(aes(x = distance_bin)) +
  geom_pointrange(aes(y = 100*prop_fpQTLs, 
                      ymin = 100*prop_fpQTLs_95low, 
                      ymax = 100*prop_fpQTLs_95high)) +
  ylim(0, NA) +
  ylab("% of variants that are fpQTLs") +
  xlab("Distance to TSS") +
  # scale_size_continuous(limits = c(0,NA), range = c(0.5,2), 
  #                       name = "Number\nof SNPs\n(thousands)") +
  scale_x_discrete(labels = c("<1 kb", "1-10 kb", "10-25 kb", 
                              "25-50 kb", "50-100 kb", ">100 kb")) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 30, vjust = 1, hjust=1))
ggsave("figures/%s/%s_tss_distance_vs_fpQTL_proportion.png" %>% sprintf(FP_METHOD, FP_METHOD), 
       width = 3, height = 3, units = "in", dpi = 600)
```


```{r}
# regression_results %>%
#   arrange(fpQTL) %>%
#   ggplot(aes(x = distance_to_nearest_tss/1000, y = abs(beta), color = fpQTL)) +
#   geom_point(size = 1) +
#   xlim(0, 100) +
#   xlab("Distance to nearest TSS (kb)") +
#   ylab("Absolute effect size") +
#   scale_color_manual(values = c("black", "red")) + 
#   theme_classic()
# 
# ggsave(sprintf("figures/%s_tss_distance_vs_effect_size.png", FP_METHOD),
#        width = 6, height = 4, units = "in")
# 
# regression_results %>%
#   arrange(fpQTL) %>%
#   ggplot(aes(x = distance_to_nearest_tss/1000, y = -log10(pval), color = fpQTL)) +
#   geom_point(size = 1) +
#   xlim(0, 100) +
#   xlab("Distance to nearest TSS (kb)") +
#   ylab(bquote(-log[10](p))) +
#   scale_color_manual(values = c("black", "red")) + 
#   theme_classic() 
# 
# ggsave(sprintf("figures/%s_tss_distance_vs_significance.png", FP_METHOD),
#        width = 6, height = 4, units = "in")
```


Are fpQTLs more likely than other variants to be within 1 kb of a TSS (in a promoter)?
```{r}
regression_results %>%
  count(fpQTL, distance_to_nearest_tss <= 1000) %>%
  pull(n) %>%
  matrix(nrow = 2, dimnames = list(
    c("TSS distal", "TSS proximal"),
    c("non-fpQTL", "fpQTL")
  )) ->
  m

m  %>% addmargins()
t(apply(m  %>% addmargins() ,1, function(x) x/sum(x))) * 2 * 100
apply(m  %>% addmargins() ,2, function(x) x/sum(x)) * 2 * 100
fisher.test(m)

```

Do fpQTLs in promoters have larger effect sizes than other fpQTLs?
```{r fig.width=3, fig.height=3}
regression_results %>%
  filter(fpQTL) %>%
  ggplot(aes(x = promoter, y = abs(beta))) +
  geom_boxplot(aes(fill = promoter), outlier.shape = NA) +
  geom_quasirandom(size = 0.6, width = 0.4) +
  xlab("fpQTL type") +
  scale_x_discrete(labels = c("distal (> 1 kb)", "promoter")) +
  ylab("Absolute effect size") +
  scale_fill_manual(guide = "none", values = c("#3b528b", "orange")) +
  stat_compare_means(aes(label = paste0("Wilcoxon,\np = ", after_stat(p.format))), label.x = 1.5, label.y=1) +
  theme_classic()
ggsave("figures/%s/%s_promoter_vs_distal_effect_size.png" %>% sprintf(FP_METHOD, FP_METHOD),
       width = 3, height = 3, units = "in", dpi = 600)

regression_results %>%
  filter(fpQTL) %>%
  ggplot(aes(x = promoter, y = abs(linear_beta))) +
  geom_boxplot(aes(fill = promoter), outlier.shape = NA) +
  geom_quasirandom(size = 0.6, width = 0.4) +
  xlab("fpQTL type") +
  geom_text(aes(label = "n = %d" %>% sprintf(..count..), color = promoter), y = 0.30, stat = "count") +
  scale_x_discrete(labels = c("distal (> 1 kb)", "promoter")) +
  ylab("Absolute lienar effect size") +
  scale_fill_manual(guide = "none", values = c("#3b528b", "orange")) +
  scale_color_manual(guide = "none", values = c("#3b528b", "orange")) +
  stat_compare_means(aes(label = paste0("Wilcoxon,\np = ", after_stat(p.format))), label.x = 1.5, label.y=0.2) +
  theme_classic()
ggsave("figures/%s/%s_promoter_vs_distal_linear_effect_size.png" %>% sprintf(FP_METHOD, FP_METHOD),
       width = 3, height = 3, units = "in", dpi = 600)
```

```{r}
effect_size.promoter <- regression_results %>% filter(fpQTL, promoter) %>% pull(beta) %>% abs()
effect_size.distal <- regression_results %>% filter(fpQTL, !promoter) %>% pull(beta) %>% abs()
wilcox.test(effect_size.promoter, effect_size.distal)
```



For fpQTLs, is there a relationship between effect size and distance?
```{r fig.width=3.75, fig.height=3}
regression_results %>%
  filter(fpQTL) %>%
  ggplot(aes(x = distance_to_nearest_tss/1000, y = abs(beta))) +
  geom_point(size = 1.2) +
  xlim(0, 1000) +
  stat_cor(method = "spearman", label.x = 25)+
  xlab("Distance to nearest TSS (kb)") +
  ylab("Absolute effect size") +
  theme_classic()
ggsave("figures/%s/%s_distance_vs_effect_size_fpQTLs.png" %>% sprintf(FP_METHOD, FP_METHOD),
       width = 5, height = 4, units = "in", dpi = 300)

regression_results %>%
  filter(fpQTL) %>%
  ggplot(aes(x = distance_to_nearest_tss/1000, y = abs(linear_beta))) +
  geom_point(size = 1.2) +
  xlim(0, 1000) +
  stat_cor(method = "spearman", label.x = 25)+
  xlab("Distance to nearest TSS (kb)") +
  ylab("Absolute linear effect size") +
  theme_classic()
ggsave("figures/%s/%s_distance_vs_linear_effect_size_fpQTLs.png" %>% sprintf(FP_METHOD, FP_METHOD),
       width = 5, height = 4, units = "in", dpi = 300)
```

# Version 2

<!-- What if we redo this analysis, but only consider TSS of genes that are expressed in liver? -->
<!-- ```{r} -->
<!-- liver_GTEx_median_tpm <- read.delim(paste0(ROOT_DIR, "raw_data/GTEx/GTEx_v8_liver_median_tpm.txt")) -->

<!-- liver_GTEx_median_tpm %>% -->
<!--   filter(Liver > 0) %>% -->
<!--   pull(Description) %>% -->
<!--   toupper() -> -->
<!--   liver_expressed_genes -->
<!-- ``` -->

<!-- ```{r} -->
<!-- refSeqCurated %>% -->
<!--   mutate(name2 = toupper(name2)) %>% -->
<!--   filter(name2 %in% liver_expressed_genes) -> -->
<!--   refSeq_liver_expressed -->

<!-- tss.range.liver_expressed <-  -->
<!--   GRanges(seqnames = refSeq_liver_expressed$chrom,  -->
<!--           ranges = IRanges(start = refSeq_liver_expressed$tx,  -->
<!--                            end = refSeq_liver_expressed$tx, -->
<!--                            names = refSeq_liver_expressed$name2)) -->
<!-- ``` -->

<!-- ```{r} -->
<!-- result <- distanceToNearest(variant.range, tss.range.liver_expressed) -->
<!-- nearest_index <- subjectHits(result) -->
<!-- regression_results$nearest_tss_liver_expressed <- refSeq_liver_expressed$name2[nearest_index] -->
<!-- regression_results$distance_to_nearest_tss_liver_expressed <- mcols(result)$distance -->
<!-- ``` -->

<!-- ```{r} -->
<!-- regression_results %>% -->
<!--   filter(fpQTL) %>% -->
<!--   ggplot(aes(x = distance_to_nearest_tss_liver_expressed/1000, y = abs(beta))) + -->
<!--   geom_point() + -->
<!--   xlim(0, 100000/1000) + -->
<!--   stat_cor(method = "spearman", label.x.npc = "middle")+ -->
<!--   xlab("Distance to nearest TSS (kb)") + -->
<!--   ylab("Absolute effect size") + -->
<!--   theme_classic() -->
<!-- ``` -->


# ENCODE rE2G

```{r}
rE2G_DIR <- paste0(ROOT_DIR, "raw_data/ENCODE/rE2G_thresholded_EG_links/")
ENCODE_rE2G_liver <- rbind(
  read.delim(paste0(rE2G_DIR, "liver_ENCFF216HEG.bed.gz") %>% gzfile()),
  read.delim(paste0(rE2G_DIR, "liver_right_lobe_ENCFF038MLV.bed.gz") %>% gzfile()),
  read.delim(paste0(rE2G_DIR, "liver_left_lobe_ENCFF059DHV.bed.gz") %>% gzfile())
)

ENCODE_rE2G_liver <- ENCODE_rE2G_liver %>%
  dplyr::rename(chrom = X.chr) %>%
  select(chrom, start, end, TargetGene)

rE2G.range <- GRanges(seqnames = ENCODE_rE2G_liver$chrom, 
                      ranges = IRanges(start = ENCODE_rE2G_liver$start, 
                                      end = ENCODE_rE2G_liver$end,
                                      names = ENCODE_rE2G_liver$TargetGene))
```


```{r}
overlaps <- GenomicRanges::findOverlaps(variant.range, rE2G.range, select = "arbitrary")
overlaps <- ifelse(is.na(overlaps), 0, 1)
regression_results$ENCODE_rE2G <- overlaps
```

# ENCODE cCRE
```{r}
ENCODE_cCRE_combined <- read.delim(paste0(ROOT_DIR, "raw_data/ENCODE/encodeCcreCombined_hg38.tsv"))

ENCODE_cCRE <- list()
ENCODE_cCRE.range <- list()
for (label in c("prom", "enhP", "enhD", "K4m3", "CTCF")) {
  ENCODE_cCRE[[label]] <- ENCODE_cCRE_combined %>% 
    filter(ucscLabel == label) %>% 
    select(X.chrom, chromStart, chromEnd)
  
  ENCODE_cCRE.range[[label]] <- GRanges(seqnames = ENCODE_cCRE[[label]]$X.chrom, 
                                        ranges = IRanges(start = ENCODE_cCRE[[label]]$chromStart,
                                                         end = ENCODE_cCRE[[label]]$chromEnd))
}
```

```{r}
for (label in c("prom", "enhP", "enhD", "K4m3", "CTCF")) {
  colname <- sprintf("cCRE_%s", label)
  
  overlaps <- GenomicRanges::findOverlaps(variant.range, ENCODE_cCRE.range[[label]], select = "arbitrary")
  overlaps <- ifelse(is.na(overlaps), 0, 1)
  regression_results[[colname]] <- overlaps
}
```



# eQTLs
```{r}
# eqtl_sig_pairs <- read.delim(paste0(ROOT_DIR, "raw_data/GTEx/GTEx_Analysis_v8_eQTL/Liver.v8.signif_variant_gene_pairs.txt"))
# gtex_all_vars_table <- read.delim(paste0(ROOT_DIR, "raw_data/GTEx/GTEx_Analysis_2017-06-05_v8_WholeGenomeSeq_838Indiv_Analysis_Freeze.lookup_table.location.txt"))
# 
# eQTL_SNPs <- eqtl_sig_pairs %>%
#   separate(variant_id, into = c("chrom", "pos", NA, NA, NA), sep = "_") %>%
#   mutate(variant_location = sprintf("%s:%s", chrom, pos)) %>%
#   pull(variant_location)
# 
# GTEx_tested_SNPs <- gtex_all_vars_table %>%
#   mutate(variant_location = sprintf("%s:%d", chr, variant_pos)) %>%
#   pull(variant_location)
# 
# eQTL_SNPs %>% saveRDS(paste0(ROOT_DIR, "raw_data/GTEx/GTEx_v8_liver_eQTLs.Rds"))
# GTEx_tested_SNPs %>% saveRDS(paste0(ROOT_DIR, "raw_data/GTEx/GTEx_v8_tested_SNPs.Rds"))
```

```{r}
eQTL_SNPs <- readRDS(paste0(ROOT_DIR, "raw_data/GTEx/GTEx_v8_liver_eQTLs.Rds"))
GTEx_tested_SNPs <- readRDS(paste0(ROOT_DIR, "raw_data/GTEx/GTEx_v8_tested_SNPs.Rds"))
```

```{r}
all_variants_eQTL <- regression_results %>%
  select(variant_chrom, variant_pos, fpQTL) %>%
  mutate(variant_location = sprintf("%s:%d", variant_chrom, variant_pos)) %>%
  filter(variant_location %in% GTEx_tested_SNPs) %>%
  mutate(eQTL = variant_location %in% eQTL_SNPs)
      
sprintf("Number of variants tested in GTEx = %d/%d", nrow(all_variants_eQTL), nrow(regression_results))
```

# Odds ratio tables
```{r}
OR_test <- function(table2x2, name) {
  ft <- fisher.test(table2x2)
  print(ft)
  
  df <- data.frame(
    annot_name=name,
    snps_neither = table2x2[1,1],
    snps_fpQTL_only = table2x2[1,2],
    snps_annot_only = table2x2[2,1],
    snps_annot_and_fpQTL = table2x2[2,2],
    odds_ratio = ft$estimate,
    odds_ratio_95_low = ft$conf.int[1],
    odds_ratio_95_high = ft$conf.int[2],
    fishers_p = ft$p.value,
    row.names = NULL
  )
  
  return(df)
}

enrichment_ORs <- data.frame()

# ENCODE rE2G
table2x2 <- table(regression_results$fpQTL, regression_results$ENCODE_rE2G)
rownames(table2x2) <- c("non-enhancer", "ENCODE_rE2G")
colnames(table2x2) <- c("non-fpQTL", "fpQTL")
table2x2
df <- OR_test(table2x2, "ENCODE-rE2G")
enrichment_ORs <- rbind(enrichment_ORs, df)

# ENCODE cCREs
for (label in c("prom", "enhP", "enhD", "K4m3", "CTCF")) {
  colname <- sprintf("cCRE_%s", label)
  
  table2x2 <- table(regression_results$fpQTL, regression_results[[colname]])
  rownames(table2x2) <- c("NA", label)
  colnames(table2x2) <- c("non-fpQTL", "fpQTL")
  print(table2x2)
  df <- OR_test(table2x2, "ENCODE cCRE\n(%s)" %>% sprintf(label))
  enrichment_ORs <- rbind(enrichment_ORs, df)
  
}

# Liver eQTLs
table2x2 <- table(all_variants_eQTL$fpQTL, all_variants_eQTL$eQTL)
rownames(table2x2) <- c("non-eQTL", "Liver eQTL")
colnames(table2x2) <- c("non-fpQTL", "fpQTL")
table2x2
df <- OR_test(table2x2, "Liver eQTL\n(GTEx)")
enrichment_ORs <- rbind(enrichment_ORs, df)

# TSS proximity
table2x2 <- table(regression_results$fpQTL, regression_results$distance_to_nearest_tss <= 500)
rownames(table2x2) <- c("distal", "proximal")
colnames(table2x2) <- c("non-fpQTL", "fpQTL")
table2x2
df <- OR_test(table2x2, "TSS-proximal\n(<500 bp)")
enrichment_ORs <- rbind(enrichment_ORs, df)
```
# MANUSCRIPT - Figure 2B
```{r fig.height=4, fig.width=5}
enrichment_ORs %>%
  ggplot(aes(x = odds_ratio, y = annot_name, color = -log10(fishers_p))) + 
  geom_vline(xintercept = 1) +
  geom_pointrange(aes(xmin = odds_ratio_95_low, 
                      xmax = odds_ratio_95_high), size = 1) +
  geom_point(data = . %>% filter(fishers_p < 0.05), shape = "*", color = "white", size = 4) +
  geom_text(aes(label = snps_annot_and_fpQTL), color = "black", nudge_y = 0.4) +
  scale_color_viridis(option = "plasma", direction = -1, name = bquote(-log[10](p)), end = 0.8) +
  scale_x_log10() +
  xlab("Odds ratio") +
  ylab("Functional annotation") +
  theme_bw() +
  theme(plot.title = element_text(size = 10))

ggsave("figures/%s/%s_OR_enrichment.png" %>% sprintf(FP_METHOD, FP_METHOD), 
       width = 5, height = 4, units = "in")
```

# Minor allele frequency
# MANUSCRUPT - Supplemental figures
```{r fig.width=3, fig.height=3}
library(scales)
regression_results %>% 
  ggplot(aes(x = maf)) +
  geom_histogram(bins = 10, fill = "black") +
  xlab("Minor Allele Frequency") +
  scale_y_continuous(labels = label_comma()) +
  theme_classic()
ggsave("figures/%s/%s_MAF_all.png" %>% sprintf(FP_METHOD, FP_METHOD), 
       width = 3, height = 3, units = "in")
```

# MANUSCRUPT - Supplemental figures
```{r fig.width=3.7, fig.height=3}
regression_results %>% 
  filter(fpQTL) %>%
  mutate(chr17_inv = variant_chrom == "chr17" & variant_pos > 4.55e7 & variant_pos < 4.64e7) %>%
  ggplot(aes(x = maf, fill = !chr17_inv)) +
  geom_histogram(bins = 10) +
  scale_fill_manual(name = "", values = c("red", "black"), 
                    labels = c("chr17\ninv.", "")) +
  xlab("Minor Allele Frequency") +
  theme_classic()
ggsave("figures/%s/%s_MAF_fpQTLs.png" %>% sprintf(FP_METHOD, FP_METHOD), 
       width = 3.7, height = 3, units = "in")
```



