```{r setup, message=FALSE, warning=FALSE}
library(tidyverse)
library(GenomicRanges)
ROOT_DIR <- "/mnt/isilon/sfgi/dudekm/"
```

(Do for every footprint method)

For every motif:

- If the motif TF has ChIP-seq data:
  - only consider motifs which overlap a ChIP-seq peak (in liver, hepatocyte, or HepG2)

Annotate the motif_resuls table with fpQTL status and beta
- store the *matched* variants in a table containing all motifs

Get the following info:
- How many variants are fpQTLs?
- How many variants have a match (ref or alt)?
- How many variants are fpQTLs AND have a match?
- Odds ratio/fisher's test p-value: do fpQTLs significantly overlap matches?
- Of variants that have an fpQTL and a match:
  - How many are "concordant" (fpQTL beta and delta motif score have same sign)?
  
and save the annotated *matched* variants to motif_match_results
- separate file for every FP method

```{r}
FP_METHOD <- "PRINT_beta_cloglog"
MOTIF_THRESHOLD <- "5e-4"
```


```{r}
chip_peaks <- read.delim(paste0(ROOT_DIR, "raw_data/ChIP-seq/tf_clusters_liver_ENCODE/new_hg38/hg38_tf_clusters_liver_labelled.txt"))

chip_peaks <- chip_peaks %>%
  mutate(name = toupper(name))

chip_peaks %>%
  count(name) %>%
  arrange(-n) %>%
  pull(name) ->
  ChIP_TF_names
```


```{r}
# Get a list of variants used for motif_results (smaller than the total list)
sample_results <- readRDS(paste0("motif_results/p=", MOTIF_THRESHOLD, "/MA1489.1_FOXN3.Rds"))
```

```{r}
# Get positions of those variants
variant_info <- readRDS("../regression/variant_info.Rds")

variant_info <- variant_info %>%
  filter(variant_id %in% sample_results$variant_id)

variant_range <- GRanges(seqnames = variant_info$variant_chrom,
                         ranges = IRanges(start = variant_info$variant_pos, 
                                          end = variant_info$variant_pos))
```


```{r}
motif_result_files <- list.files(paste0("motif_results/",MOTIF_THRESHOLD,"/"), 
                                 "*.Rds", full.names = TRUE)
motif_ids <- str_match(motif_result_files, ".*/(.*?)\\.Rds")[,2]
```

```{r}
motif_info <-
  data.frame(
    motif_id = motif_ids
  ) %>%
  separate(col = "motif_id", into = c("motif_id2", "TF_name"), 
           sep = "_", remove = FALSE) %>%
  mutate(TF_name = toupper(TF_name)) %>%
  separate_rows(TF_name, sep = "::") %>%
  mutate(has_chip_data = TF_name %in% ChIP_TF_names)

motif_info %>%
  write.table("motif_summary/motif_info.txt",
              quote = FALSE, row.names = FALSE, sep = "\t")

motif_info %>%
  filter(has_chip_data)
```


```{r}
regression_results <- readRDS(paste0("../regression/FP_methods/", FP_METHOD, "/regression_results/fp_score_covariates_genotype_regression.Rds"))

if (grepl("local_cutsites_geq10", FP_METHOD, fixed = TRUE)) {
  Q_THRESHOLD <- 0.10
} else {
  Q_THRESHOLD <- 0.05
}

regression_results %>%
  # Only consider variants we have results for
  filter(variant_id %in% sample_results$variant_id) %>% 
  mutate(fpQTL = ST_qval < Q_THRESHOLD) ->
  regression_results

#Ensure variant order is correct
all(regression_results$variant_id == sample_results$variant_id)
```

Testing
```{r}
i <- 4
motif_result_file <- motif_result_files[i]
motif_id_ <- motif_ids[i]
cat(motif_id, "\n", sep = "")


motif_result <- readRDS(motif_result_file)

TFs_with_chip_data <- motif_info %>% 
  filter(motif_id == motif_id_, has_chip_data) %>%
  pull(TF_name)

if (length(TFs_with_chip_data) == 0) {TFs_with_chip_data <- NA}

# Filter for variants within 10 bp of a ChIP peak
if (!is.na(TFs_with_chip_data)) {
  TF_chip_peaks <- chip_peaks %>%
    filter(name %in% TFs_with_chip_data)
  
  TF_chip_peaks_range <- GRanges(seqnames = TF_chip_peaks$chrom,
                                 ranges = IRanges(start = TF_chip_peaks$chromStart, 
                                          end = TF_chip_peaks$chromEnd))
  
  overlaps <- GenomicRanges::findOverlaps(variant_range, TF_chip_peaks_range, maxgap = 10)
  
  motif_result <- motif_result[queryHits(overlaps), ]
  
}

for (i in c(-1, 5, 10, 20, 50, 100, 200, 300)) {
  overlaps <- GenomicRanges::findOverlaps(variant_range, TF_chip_peaks_range, maxgap = i)
  print(i)
  motif_result_chip_only <- motif_result[queryHits(overlaps), ]
  print(sum(motif_result_chip_only$ref_match | motif_result_chip_only$alt_match))
}

```




```{r}
all_matches <- data.frame()
motif_summary <- data.frame()

for (i in seq_along(motif_result_files)) {
  motif_result_file <- motif_result_files[i]
  motif_id <- motif_ids[i]
  cat(motif_id, "\n", sep = "")
  
  
  motif_result <- readRDS(motif_result_file)
  
  motif_result %>%
    mutate(
      delta_score = alt_score - ref_score,
      match = ref_match | alt_match,
      beta = regression_results$beta,
      fpQTL = regression_results$fpQTL
    ) ->
    motif_result
  
  
  motif_result %>%
    filter(match) %>%
    mutate(motif_id = motif_id) %>%
    select(-match) ->
    motif_result_matches
  
  all_matches <- rbind(all_matches, motif_result_matches)
  
  motif_result %>%
    count(match, fpQTL) %>%
    complete(match, fpQTL, fill = list(n = 0L)) %>%
    pull(n) %>%
    matrix(nrow = 2,
           dimnames = list(
             c("non_fpQTL", "fpQTL"),
             c("no_match", "match")
           )) ->
    m
  
  num_match_fpQTL = m["fpQTL", "match"]
  
  # Test - are fpQTLs enriched for motif matches?
  ft <- fisher.test(m)
  
  # How many fpQTL matches are concordant?
  # i.e., the fpQTL effect size and the motif score change are in the same direction
  motif_result %>%
    filter(match & fpQTL) %>%
    filter(sign(beta) == sign(delta_score)) %>%
    nrow() ->
    num_concordant
 
  
  # Test - are fpQTLs more likely to be concordant than at random (p = 0.5)?
  bt <- binom.test(num_concordant, num_match_fpQTL, p = 0.5)
  
  motif_summary_row <- data.frame(
    motif_id = motif_id,
    num_matches = sum(m[,"match"]),
    num_fpQTLs = sum(m["fpQTL",]),
    num_match_fpQTL = num_match_fpQTL,
    num_match_fpQTL_concordant = num_concordant,
    fpQTL_match_OR = ft$estimate,
    fpQTL_match_OR_95_low = ft$conf.int[1],
    fpQTL_match_OR_95_high = ft$conf.int[2],
    fpQTL_match_p = ft$p.value,
    concordant_prob = bt$estimate,
    concordant_prob_95_low = bt$conf.int[1],
    concordant_prob_95_high = bt$conf.int[2],
    concordant_p = bt$p.value,
    row.names = NULL
  )
  
  motif_summary <- rbind(motif_summary, motif_summary_row)
}

all_matches %>% saveRDS(paste0("motif_match_results/", FP_METHOD, "_motif_matches.Rds"))
motif_summary %>%
  write.table(paste0("motif_summary/", FP_METHOD, "_motif_summary.txt"),
              sep = "\t", row.names = FALSE, quote = FALSE)
```




